### PR TITLE
Add new check "implicit-flag-alias"

### LINKED ISSUE #8102: Feature request: implicit-flag-overlap
### Current problem

The following behaviour seemed confusing:

```py
>>> from enum import IntFlag
>>> class TestFlag(IntFlag):
...     A = 1
...     B = 2
...     C = 3
... 
>>> print([i for i in TestFlag])
[<TestFlag.A: 1>, <TestFlag.B: 2>]
```

(note that this is an `IntFlag`, not an `IntEnum`)

### Desired solution

The reason this happens is that `Flag` types are designed for setwise operations (bitwise operations, in the case of `IntFlag`), and, because `C` can be composed of `A` and `B`...

```py
>>> int(TestFlag.A | TestFlag.B)
3
>>> int(TestFlag.C)
3
```

...internally, `IntFlag` considers `C` to be an alias -- and doesn't include it when iterating over the members of the flag.

A `pylint` message something along the lines of `implicit-flag-overlap` could be emitted by a relevant checker when scanning the class declaration.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jayaddison**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

This changeset implements an `implicit-flag-alias` check that inspects classes derived from `enum.Flag` and looks for member values that share common bit positions.

The commented sample code below may help explain:

```py
from enum import IntFlag

class Test(IntFlag):

    # valid cases:

    A = 1
    B = 2
    C = 2          # C is an alias (alternative name) for B, and is not highlighted by the check
    D = A | B      # D shares bit positions with A and B, but uses bitwise-operator syntax and is not highlighted

    # implicit-flag-alias cases:

    E = 3          # The value of E shares bit positions with A and B, so the check will highlight it
```

Closes #8102.

### CODE REVIEW

File: pylint/checkers/classes/class_checker.py
Comment (jayaddison) [inline pylint/checkers/classes/class_checker.py:932]: Note: there's an assumption here that subclasses of `enum.Flag` will implement addition/sum operations - I don't know whether that's reasonable (the check could be adjusted to only inspect subclasses of `enum.IntFlag`).

File: pylint/checkers/classes/class_checker.py
Comment (clavedeluna) [inline pylint/checkers/classes/class_checker.py:914]: `visit_classdef` is quite bloated already, would you mind moving the new code to a `check_flag` (better name maybe) method?

File: pylint/checkers/classes/class_checker.py
Comment (clavedeluna) [inline pylint/checkers/classes/class_checker.py:939]: add_message should also have `confidence=INFERENCE` (and update functional test .txt output accordingly)

File: pylint/checkers/classes/class_checker.py
Comment (jayaddison) [reply pylint/checkers/classes/class_checker.py:914]: Would it be OK to relocate the `enum.Enum` code at the same time (`_check_enum_bases`, or similar)?  Reasoning: it might be possible to de-duplicate some of the conditional logic after doing that.

Comment (clavedeluna) [reply pylint/checkers/classes/class_checker.py:914]: yeah for sure!

**Comment by mbyrnepr2** [comment]
I feel the naming, especially the word “overlap” isn’t intuitive. I believe we are talking about warning when a flag member alias is defined, so what about flag-member-alias-defined, or  something similar, as a more direct message?

**Comment by jayaddison** [comment]
Thanks for the feedback @mbyrnepr2 - this check _excludes_ aliases (if `A = 1` and `B = 1` are two members defined in an enum, no warning will be emitted for either of them).

**Comment by mbyrnepr2** [comment]
> Thanks for the feedback @mbyrnepr2 - this check _excludes_ aliases (if `A = 1` and `B = 1` are two members defined in an enum, no warning will be emitted for either of them).

I was referring to the initial code example in the issue and in the functional test where an IntFlag member is excluded from the iteration over the Enum when it can be the result of a bitwise or of two other members.

File: pylint/checkers/classes/class_checker.py
Comment (jayaddison) [reply pylint/checkers/classes/class_checker.py:914]: I'm experimenting with this a bit locally; it's making me wonder whether it'd be possible (and more efficient) to use the existing tree `visit_` handlers to achieve the same check.

**Comment by jayaddison** [comment]
> where an IntFlag member is excluded from the iteration over the Enum when it can be the result of a bitwise or of two other members

Ok, my apologies - now I understand, I think - my use of the term _alias_ has been out-of-alignment with the documentation (in `A = 1, B = 2, C = 3`, `C` is considered an alias).

Previously I thought that _alias_ referred purely to secondary attribute names for a single value (as in my `A = 1, B = 1` example).

Spending some time to think about renaming the check.. initially `implicit-flag-alias` springs to mind (trying to keep the name concise and combining your correct use of the terminology).

File: pylint/checkers/classes/class_checker.py
Comment (DanielNoord) [inline pylint/checkers/classes/class_checker.py:918]: I think it would be really nice to use the `args` argument of `add_message` to tell the user a little bit about which flag is overlapping (its name) and potentially also with which other flag.

**Review by DanielNoord — COMMENTED** [review]
Sorry for another comment instead of a well structured review but the code already looks really good and this is the only major thing I see now.

Really nice check btw! Never would have thought of this myself and I can see this preventing some people from making mistakes!

**Comment by jayaddison** [comment]
(and I'm also reading about some of the [`EnumCheck` decorator functionality](https://docs.python.org/3/library/enum.html#enum.EnumCheck))

**Comment by DanielNoord** [comment]
I think the primer is also showing that this causes a `Fatal` error somewhere. Our documentation has instructions about running the primer locally, which should help debugging this.

File: pylint/checkers/classes/class_checker.py
Comment (jayaddison) [inline pylint/checkers/classes/class_checker.py:943]: It's not directly relevant to this pull request, but I wonder whether moving this `baseclass-is-object` check before the inference steps, with a `continue` statement, might save some time:

```python
        for base in node.bases:
            if getattr(base, 'name', None) == object.__name__:
                self.add_message(
                    "useless-object-inheritance", args=node.name, node=node
                )
                continue
```

File: pylint/checkers/classes/class_checker.py
Comment (jayaddison) [reply pylint/checkers/classes/class_checker.py]: (maybe: for now, it seems simpler and likely more maintainable to keep the logic grouped textually)

File: pylint/checkers/classes/class_checker.py
Comment (jayaddison) [reply pylint/checkers/classes/class_checker.py]: That sounds good - yep, I'll give this a try :+1:

Comment (jayaddison) [reply pylint/checkers/classes/class_checker.py]: Partially-implemented: the message includes information about the flag member where an implicit alias was detected, but doesn't match this up with the corresponding flag(s) that conflict.  I couldn't think of a way to achieve that while also maintaining linear time and space bounding.

File: tests/functional/i/implicit/implicit_flag_alias.txt
Comment (jayaddison) [inline tests/functional/i/implicit/implicit_flag_alias.txt:3]: Suggested improvements for the message phrasing here would be great, I think it reads a bit awkwardly.

File: pylint/checkers/classes/class_checker.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:529]:
```suggestion
        "Flag member <%s.%s: %d> implicitely hide <%s.%s: %d> as %s|%s=%s.",
```

**Review by Pierre-Sassoulas — COMMENTED** [review]
Looks pretty good already !

**Comment by mbyrnepr2** [comment]
> > where an IntFlag member is excluded from the iteration over the Enum when it can be the result of a bitwise or of two other members
> 
> Ok, my apologies - now I understand, I think - my use of the term _alias_ has been out-of-alignment with the documentation (in `A = 1, B = 2, C = 3`, `C` is considered an alias).
> 
> Previously I thought that _alias_ referred purely to secondary attribute names for a single value (as in my `A = 1, B = 1` example).
> 
> Spending some time to think about renaming the check.. initially `implicit-flag-alias` springs to mind (trying to keep the name concise and combining your correct use of the terminology).

My initial impression was that `alias` would be good to have in there somewhere because that is the [Python documentation's naming](https://docs.python.org/3/library/enum.html#enum.Flag) for it (see section on `__iter__`.

I would say don't change the naming unless someone else also finds that the suggestion makes sense - let's not be led by my impressions but validate with someone else first :D

File: pylint/checkers/classes/class_checker.py
Comment (mbyrnepr2) [reply pylint/checkers/classes/class_checker.py:529]:
```suggestion
        "Flag member <%s.%s: %d> implicitly hides <%s.%s: %d> as %s|%s=%s.",
```

(This is my new specialty :D)

File: pylint/checkers/classes/class_checker.py
Comment (jayaddison) [reply pylint/checkers/classes/class_checker.py]: This is now fully-[implemented](https://github.com/PyCQA/pylint/pull/8105/commits/ac21d3149c2b67365249c16ac5853309de93cea3), hopefully in a reasonably-efficient way.

File: pylint/checkers/classes/class_checker.py
Comment (jayaddison) [reply pylint/checkers/classes/class_checker.py]: Adapted these suggested changes slightly - I found it easier to understand the cause when shown the intersection (`A & B`) rather than the union (`A | B`) of bitmasks where this issue can occur.

**Comment by jayaddison** [comment]
>  My understanding is that this checker is saying: `Color.BLUE` is an alias for `Color.GREEN & Color.BLUE` and that it may not be necessary to define `Color.BLUE` at all for that reason.

Mm, not quite - it's currently saying that the symbol for `Color.BLUE` and the symbol for `Color.GREEN` share some common elements -- in fact, in the example, `Color.BLUE` contains all of the symbols from `Color.GREEN` -- which may not be what the author intends.

(if they do, they can declare `BLUE = GREEN | RED` and the warning goes away)

>  I think the wording could be changed to something like `Flag member Color.BLUE: 3 is an alias of Color.RED | Color.GREEN` as a simpler wording

That's a nice idea, although there are two problems I can think of:

- I'm not sure it's possible to implement efficiently (can you suggest an algorithm?)
- I don't know what output you'd expect for a class containing `A = 1, B = 3, C = 5` - can you clarify?

**Comment by jayaddison** [comment]
No problem, thanks @mbyrnepr2 - I'm still finding it slightly difficult to describe concisely what the check does (even after writing it), so that's probably a sign that the documentation and messaging could be improved further.

File: pylint/checkers/classes/class_checker.py
Comment (jayaddison) [inline pylint/checkers/classes/class_checker.py:915]: In hindsight, this is slightly odd: why only report the maximum overlapping value when multiple items share a common bit position?

(todo: more test coverage to examine this)

File: tests/functional/i/implicit/implicit_flag_alias.py
Comment (jayaddison) [inline tests/functional/i/implicit/implicit_flag_alias.py:30]: TODO: Add some (intentional) aliases - there are some edge cases to handle around those.

For example:

```
G = 17
```

File: pylint/checkers/classes/class_checker.py
Comment (jayaddison) [reply pylint/checkers/classes/class_checker.py]: There was indeed a problem here, and it's now resolved (with test coverage).

Basically, if `1, 5, 9` (for example) were declared as flag values, then the min/max logic that existed here previously would have resulting in the `9` as having a reported overlap, but would not have reported about the `5`.

In practice anyone who noticed that by using `pylint` would probably have run the check again after fixing the `9`, and _then_ been told about the `5`, but that'd be kinda an annoying workflow.

I think the pull request is ready for review again.

File: pylint/checkers/classes/class_checker.py
Comment (jayaddison) [reply pylint/checkers/classes/class_checker.py:959]: (in practice: no, this doesn't seem worthwhile to move or adjust at the moment; removing the check entirely doesn't result in a significant difference in the timings produced by `tests/benchmark/test_baseline_benchmarks.py` locally, so I don't think there's a huge opportunity for improvement here)

File: pylint/checkers/classes/class_checker.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:529]: Great way to make the code more maintanable, I wonder if we should doit everywhere.

**Review by Pierre-Sassoulas — APPROVED** [review]
LGTM, great checker ! Could go in 2.16.0 maybe.

**Comment by jayaddison** [comment]
Thanks again @Pierre-Sassoulas!  There's no hurry as far as I'm concerned, so I'll leave it up to you as maintainers.

It might be slightly easier for me to respond to any feedback for one change at a time, so I have a mild preference to wait until after 2.16 for this one.. but either way I'll follow the changes and watch for relevant tickets after release.

File: pylint/checkers/classes/class_checker.py
Comment (jayaddison) [reply pylint/checkers/classes/class_checker.py:529]: Unfortunately it's not possible to express in a `pylint` rule (because it also involves `diff` context), but I think my approach would be:

If a string is being modified by a changeset _and_ contains `%`-replacement characters, then consider also changing the template to use named variables (or f-strings) during the same change -- although with some caution about larger potential refactoring if doing so would result in line-reformatting due to `black` / `flake8` rules.

**Comment by DanielNoord** [comment]
@jayaddison You removed your repository and therefore this branch. Does that imply that you no longer want to work on this PR?

**Comment by jayaddison** [comment]
@DanielNoord argh, thanks - my mistake.  I forgot that this was still pending when deleting the repo after my contributions-alias edit (#8343) was merged.  I'm good to continue working on this PR and will re-open it.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
The documentation fail is genuine, an example need to be added :)

**Comment by jayaddison** [comment]
Okie doke - I think I'll open a second pull request (re-using the same commits from here) with an example added, since I'm not sure I'll be able to push further commits to this branch now that the repo has been deleted.

**Comment by jayaddison** [comment]
Continues in #8345.

**Pull request closed without merging**
